File: docs/01-beginner-tmb/exercises/codend.cpp
#include <TMB.hpp>
template<class Type>
  Type objective_function<Type>::operator() ()
{
  
  DATA_VECTOR(Codend);
  DATA_VECTOR(Cover);
  DATA_VECTOR(Length);
  
  PARAMETER(beta0);
  PARAMETER(beta1);
  
  Type nll;
  vector<Type> N = Codend + Cover;
  vector<Type> eta  = beta0 + beta1 * Length;
  vector<Type> p = 1/(1+exp(-eta));

  nll = -sum(dbinom(Codend, N, p, true));
  
  Type L50 = -beta0/beta1;
  REPORT(L50);
  ADREPORT(L50);
  REPORT(p);
  
  return nll;
  
}

File: docs/01-beginner-tmb/exercises/codend.R
load("docs/01-beginner-tmb/exercises/stownet.RData")

Dat <- list(
  Codend = stownet$Codend,
  Cover = stownet$Cover,
  Length = stownet$Lenclass
)

Par <- list(
  beta0 = 0,
  beta1 = 0
)
library(TMB)
compile("docs/01-beginner-tmb/exercises/codend.cpp")
dyn.load(dynlib("docs/01-beginner-tmb/exercises/codend"))

obj <- MakeADFun(Dat, Par, DLL = "codend")
opt <- nlminb(obj$par, obj$fn, obj$gr)
opt$par
opt$objective

sdr <- sdreport(obj)
sdr
summary(sdr, "report")

